File: qswitch.py
# -*- coding: utf-8 -*-
from PyQt5.QtCore import (
    QSize,
    pyqtSignal,
    pyqtProperty,
    QPropertyAnimation,
    QEasingCurve,
    QPointF,
    QParallelAnimationGroup,
    QPoint,
)
from PyQt5.QtGui import QColor, QResizeEvent, QMouseEvent
from PyQt5.QtWidgets import QWidget, QFrame, QGraphicsDropShadowEffect


class _QSwitchLayer(QFrame):
    def backgroundColor(self) -> QColor:
        return self._backgroundColor

    def setBackgroundColor(self, newValue: QColor) -> None:
        self._backgroundColor = newValue
        self._applyStyleSheet()

    def borderColor(self) -> QColor:
        return self._borderColor

    def setBorderColor(self, newValue: QColor) -> None:
        self._borderColor = newValue
        self._applyStyleSheet()

    def borderRadius(self) -> int:
        return self._borderRadius

    def setBorderRadius(self, newValue: int) -> None:
        self._borderRadius = newValue
        self._applyStyleSheet()

    backgroundColor = pyqtProperty(
        QColor, fget=backgroundColor, fset=setBackgroundColor
    )
    borderColor = pyqtProperty(QColor, fget=borderColor, fset=setBorderColor)
    borderRadius = pyqtProperty(int, fget=borderRadius, fset=setBorderRadius)
    _templateStyleSheet = '_QSwitchLayer {{ background: rgba({}, {}, {}, {}); border: {}px solid rgba({}, {}, {}, {}); border-radius: {}px; }}'

    def __init__(self, parent: QWidget = None) -> None:
        super().__init__(parent)
        self.disabledBackgroundColor = self.palette().color(
            self.backgroundRole()
        )
        self.disabledBorderColor = QColor(0, 0, 0, 51)
        self._backgroundColor = self.disabledBackgroundColor
        self._borderColor = self.disabledBorderColor
        self._borderRadius = 0

    def _applyStyleSheet(self) -> None:
        self.setStyleSheet(
            _QSwitchLayer._templateStyleSheet.format(
                self.backgroundColor.red(),
                self.backgroundColor.green(),
                self.backgroundColor.blue(),
                self.backgroundColor.alpha(),
                QSwitch.BORDER_LINE_WIDTH,
                self.borderColor.red(),
                self.borderColor.green(),
                self.borderColor.blue(),
                self.borderColor.alpha(),
                self.borderRadius,
            )
        )


class QSwitch(QFrame):
    ANIMATION_DURATION_MS = 400
    BORDER_LINE_WIDTH = 1
    GOLDEN_RATIO = 1.61803398875
    DECREASED_GOLDEN_RATIO = 1.38
    KNOB_BACKGROUND_COLOR = QColor(255, 255, 255, 255)
    DEFAULT_TINT_COLOR = QColor(79, 220, 116, 255)
    DEFAULT_MARGIN = 5
    MINIMUM_WIDTH = 32
    MINIMUM_HEIGHT = 20

    def isChecked(self) -> bool:
        return self._checked

    def setChecked(self, newValue: bool) -> None:
        self._checked = newValue
        self._redrawLayers(animated=True)

    def tintColor(self) -> QColor:
        return self._tintColor

    def setTintColor(self, newValue: QColor) -> None:
        self._tintColor = newValue
        self._redrawLayers()

    toggled = pyqtSignal(bool)
    checked = pyqtProperty(bool, fget=isChecked, fset=setChecked)
    tintColor = pyqtProperty(bool, fget=tintColor, fset=setTintColor)

    def __init__(
        self, tintColor: QColor = DEFAULT_TINT_COLOR, parent: QWidget = None
    ) -> None:
        super().__init__(parent)
        self._checked = False
        self._tintColor = tintColor
        self._isMouseDown = False
        self._didDrag = False
        self._isDraggingTowardsOn = False
        self._isDraggingTowardsOnDidChangeSinceLastMoveEvent = False
        self._animations = QParallelAnimationGroup(self)

        self._backgroundLayer = _QSwitchLayer(self)
        self._backgroundLayer.setObjectName('_backgroundLayer')
        self.setMinimumSize(
            QSwitch.MINIMUM_WIDTH + 2 * QSwitch.DEFAULT_MARGIN,
            QSwitch.MINIMUM_HEIGHT + 2 * QSwitch.DEFAULT_MARGIN,
        )
        self._backgroundLayer.move(
            QSwitch.DEFAULT_MARGIN, QSwitch.DEFAULT_MARGIN
        )

        self._knobLayer = QFrame(self)
        self._knobLayer.setObjectName('_knobLayer')
        _knobLayerShadowEffect = QGraphicsDropShadowEffect()
        _knobLayerShadowEffect.setColor(QColor(0, 0, 0, 77))
        _knobLayerShadowEffect.setBlurRadius(4)
        _knobLayerShadowEffect.setOffset(0, 2)
        self._knobLayer.setGraphicsEffect(_knobLayerShadowEffect)

        self._knobInsideLayer = QFrame(self)
        self._knobInsideLayer.setObjectName('_knobInsideLayer')
        _knobLayerInsideShadowEffect = QGraphicsDropShadowEffect()
        _knobLayerInsideShadowEffect.setColor(QColor(0, 0, 0, 89))
        _knobLayerInsideShadowEffect.setBlurRadius(4)
        _knobLayerInsideShadowEffect.setOffset(0, 0)
        self._knobInsideLayer.setGraphicsEffect(_knobLayerInsideShadowEffect)

        self._redrawLayers()

    def _makeBezierCurveOvershoot(self) -> QEasingCurve:
        curve = QEasingCurve(QEasingCurve.BezierSpline)
        curve.addCubicBezierSegment(
            QPointF(0.25, 1.5), QPointF(0.5, 1.0), QPointF(1.0, 1.0)
        )
        return curve

    def _makeBezierCurveDefault(self) -> QEasingCurve:
        curve = QEasingCurve(QEasingCurve.BezierSpline)
        curve.addCubicBezierSegment(
            QPointF(0.25, 0.1), QPointF(0.25, 1.0), QPointF(1.0, 1.0)
        )
        return curve

    def _redrawLayers(self, animated: bool = False) -> None:
        self._backgroundLayer.resize(
            self.width() - 2 * QSwitch.DEFAULT_MARGIN,
            self.height() - 2 * QSwitch.DEFAULT_MARGIN,
        )
        self._backgroundLayer.borderRadius = (
            int((self.height() - 2 * QSwitch.DEFAULT_MARGIN) / 2) - 1
        )

        self._animations.clear()

        if (self._didDrag and self._isDraggingTowardsOn) or (
            not self._didDrag and self.checked
        ):
            _animationBackgroundColor = QPropertyAnimation(
                self._backgroundLayer, b'backgroundColor'
            )
            _animationBackgroundColor.setDuration(
                QSwitch.ANIMATION_DURATION_MS
            )
            _animationBackgroundColor.setStartValue(
                self._backgroundLayer.backgroundColor
            )
            _animationBackgroundColor.setEndValue(self.tintColor)
            if self._didDrag:
                _animationBackgroundColor.setEasingCurve(
                    self._makeBezierCurveDefault()
                )
            else:
                _animationBackgroundColor.setEasingCurve(
                    self._makeBezierCurveOvershoot()
                )
            self._animations.addAnimation(_animationBackgroundColor)

            _animationBorderColor = QPropertyAnimation(
                self._backgroundLayer, b'borderColor'
            )
            _animationBorderColor.setDuration(QSwitch.ANIMATION_DURATION_MS)
            _animationBorderColor.setStartValue(
                self._backgroundLayer.borderColor
            )
            _animationBorderColor.setEndValue(self.tintColor)
            if self._didDrag:
                _animationBorderColor.setEasingCurve(
                    self._makeBezierCurveDefault()
                )
            else:
                _animationBorderColor.setEasingCurve(
                    self._makeBezierCurveOvershoot()
                )
            self._animations.addAnimation(_animationBorderColor)
        else:
            _animationBackgroundColor = QPropertyAnimation(
                self._backgroundLayer, b'backgroundColor'
            )
            _animationBackgroundColor.setDuration(
                QSwitch.ANIMATION_DURATION_MS
            )
            _animationBackgroundColor.setStartValue(
                self._backgroundLayer.backgroundColor
            )
            _animationBackgroundColor.setEndValue(
                self._backgroundLayer.disabledBackgroundColor
            )
            if self._didDrag:
                _animationBackgroundColor.setEasingCurve(
                    self._makeBezierCurveDefault()
                )
            else:
                _animationBackgroundColor.setEasingCurve(
                    self._makeBezierCurveOvershoot()
                )
            self._animations.addAnimation(_animationBackgroundColor)

            _animationBorderColor = QPropertyAnimation(
                self._backgroundLayer, b'borderColor'
            )
            _animationBorderColor.setDuration(QSwitch.ANIMATION_DURATION_MS)
            _animationBorderColor.setStartValue(
                self._backgroundLayer.borderColor
            )
            _animationBorderColor.setEndValue(
                self._backgroundLayer.disabledBorderColor
            )
            if self._didDrag:
                _animationBorderColor.setEasingCurve(
                    self._makeBezierCurveDefault()
                )
            else:
                _animationBorderColor.setEasingCurve(
                    self._makeBezierCurveOvershoot()
                )
            self._animations.addAnimation(_animationBorderColor)

        knobWidth = int(
            (
                self.width()
                - 2 * QSwitch.DEFAULT_MARGIN
                - 2 * QSwitch.BORDER_LINE_WIDTH
            )
            * (1 / QSwitch.GOLDEN_RATIO)
        )
        roundKnobWidth = knobWidth
        if self._isMouseDown:
            knobWidth = int(
                (
                    self.width()
                    - 2 * QSwitch.DEFAULT_MARGIN
                    - 2 * QSwitch.BORDER_LINE_WIDTH
                )
                * (1 / QSwitch.DECREASED_GOLDEN_RATIO)
            )

        if animated:
            _animationSizeKnob = QPropertyAnimation(self._knobLayer, b'size')
            _animationSizeKnob.setDuration(QSwitch.ANIMATION_DURATION_MS)
            _animationSizeKnob.setStartValue(self._knobLayer.size())
            _animationSizeKnob.setEndValue(
                QSize(
                    knobWidth,
                    self.height()
                    - 2 * QSwitch.DEFAULT_MARGIN
                    - 2 * QSwitch.BORDER_LINE_WIDTH,
                )
            )
            if not self._didDrag:
                _animationSizeKnob.setEasingCurve(
                    self._makeBezierCurveOvershoot()
                )
            else:
                _animationSizeKnob.setEasingCurve(
                    self._makeBezierCurveDefault()
                )
            self._animations.addAnimation(_animationSizeKnob)
        else:
            self._knobLayer.resize(
                knobWidth,
                self.height()
                - 2 * QSwitch.DEFAULT_MARGIN
                - 2 * QSwitch.BORDER_LINE_WIDTH,
            )

        if (not self._didDrag and not self.checked) or (
            self._didDrag and not self._isDraggingTowardsOn
        ):
            knobX = QSwitch.DEFAULT_MARGIN + QSwitch.BORDER_LINE_WIDTH
        else:
            knobX = (
                self.width()
                - knobWidth
                - QSwitch.BORDER_LINE_WIDTH
                - QSwitch.DEFAULT_MARGIN
            )

        if animated:
            _animationPosKnob = QPropertyAnimation(self._knobLayer, b'pos')
            _animationPosKnob.setDuration(QSwitch.ANIMATION_DURATION_MS)
            _animationPosKnob.setStartValue(self._knobLayer.pos())
            _animationPosKnob.setEndValue(
                QPoint(
                    knobX, QSwitch.DEFAULT_MARGIN + QSwitch.BORDER_LINE_WIDTH
                )
            )
            if not self._didDrag:
                _animationPosKnob.setEasingCurve(
                    self._makeBezierCurveOvershoot()
                )
            else:
                _animationPosKnob.setEasingCurve(
                    self._makeBezierCurveDefault()
                )
            self._animations.addAnimation(_animationPosKnob)
        else:
            self._knobLayer.move(
                knobX, QSwitch.DEFAULT_MARGIN + QSwitch.BORDER_LINE_WIDTH
            )
            self._knobLayer.setStyleSheet(
                'QFrame#_knobLayer {{ background-color: white; border-radius: {0}px; }}'.format(
                    int(self._knobLayer.height() / 2) - 1
                )
            )

        if animated:
            _animationSizeKnobInside = QPropertyAnimation(
                self._knobInsideLayer, b'size'
            )
            _animationSizeKnobInside.setDuration(QSwitch.ANIMATION_DURATION_MS)
            _animationSizeKnobInside.setStartValue(
                self._knobInsideLayer.size()
            )
            _animationSizeKnobInside.setEndValue(
                QSize(
                    knobWidth,
                    self.height()
                    - 2 * QSwitch.DEFAULT_MARGIN
                    - 2 * QSwitch.BORDER_LINE_WIDTH,
                )
            )
            if not self._didDrag:
                _animationSizeKnobInside.setEasingCurve(
                    self._makeBezierCurveOvershoot()
                )
            else:
                _animationSizeKnobInside.setEasingCurve(
                    self._makeBezierCurveDefault()
                )
            self._animations.addAnimation(_animationSizeKnobInside)
        else:
            self._knobInsideLayer.resize(
                knobWidth,
                self.height()
                - 2 * QSwitch.DEFAULT_MARGIN
                - 2 * QSwitch.BORDER_LINE_WIDTH,
            )

        if animated:
            _animationPosKnobInside = QPropertyAnimation(
                self._knobInsideLayer, b'pos'
            )
            _animationPosKnobInside.setDuration(QSwitch.ANIMATION_DURATION_MS)
            _animationPosKnobInside.setStartValue(self._knobInsideLayer.pos())
            _animationPosKnobInside.setEndValue(
                QPoint(
                    knobX, QSwitch.DEFAULT_MARGIN + QSwitch.BORDER_LINE_WIDTH
                )
            )
            if not self._didDrag:
                _animationPosKnobInside.setEasingCurve(
                    self._makeBezierCurveOvershoot()
                )
            else:
                _animationPosKnobInside.setEasingCurve(
                    self._makeBezierCurveDefault()
                )
            self._animations.addAnimation(_animationPosKnobInside)
        else:
            self._knobInsideLayer.move(
                knobX, QSwitch.DEFAULT_MARGIN + QSwitch.BORDER_LINE_WIDTH
            )
            self._knobInsideLayer.setStyleSheet(
                'QFrame#_knobInsideLayer {{ background-color: white; border-radius: {}px; }}'.format(
                    int(self._knobInsideLayer.height() / 2) - 1
                )
            )

        self._animations.start()

        self.setMaximumHeight(
            roundKnobWidth
            + 2 * QSwitch.DEFAULT_MARGIN
            + 2 * QSwitch.BORDER_LINE_WIDTH
        )

    def mousePressEvent(self, event: QMouseEvent) -> None:
        if not self.isEnabled():
            return
        self._isMouseDown = True
        self._redrawLayers(animated=True)

    def mouseMoveEvent(self, event: QMouseEvent) -> None:
        if not self._isMouseDown:
            return

        self._didDrag = True

        _oldIsDraggingTowardsOn = self._isDraggingTowardsOn
        self._isDraggingTowardsOn = (
            event.pos().x()
            >= (self.width() - 2 * QSwitch.DEFAULT_MARGIN) / 2.0
        )
        if _oldIsDraggingTowardsOn != self._isDraggingTowardsOn:
            self._isDraggingTowardsOnDidChangeSinceLastMoveEvent = True
            self._redrawLayers(animated=True)

    def mouseReleaseEvent(self, event: QMouseEvent) -> None:
        if not self.isEnabled():
            return
        self._isMouseDown = False

        if not self._didDrag:
            self.checked = not self.checked
        else:
            self.checked = self._isDraggingTowardsOn

        self._didDrag = False
        self._isDraggingTowardsOn = False

        if not self._isDraggingTowardsOnDidChangeSinceLastMoveEvent:
            self._redrawLayers(animated=True)

        self._isDraggingTowardsOnDidChangeSinceLastMoveEvent = False
        self.toggled.emit(self.checked)

    def resizeEvent(self, event: QResizeEvent) -> None:
        super().resizeEvent(event)
        self._redrawLayers()

    def sizeHint(self) -> QSize:
        hint = self._backgroundLayer.sizeHint()
        return QSize(
            hint.width() + 2 * QSwitch.DEFAULT_MARGIN,
            hint.height() + 2 * QSwitch.DEFAULT_MARGIN,
        )

File: QSwitch.cpp
#include "QSwitch.h"

#include <QDebug>
#include <QGraphicsDropShadowEffect>
#include <QPoint>
#include <QPropertyAnimation>
#include <QSize>

QSwitchLayer::QSwitchLayer(QWidget *parent)
    : QFrame(parent),
      m_backgroundColor(this->palette().color(this->backgroundRole())),
      m_borderColor(this->disabledBorderColor()),
      m_disabledBackgroundColor(
          this->palette().color(this->backgroundRole())) {}

QColor QSwitchLayer::backgroundColor() const {
    return this->m_backgroundColor;
}

void QSwitchLayer::setBackgroundColor(QColor newValue) {
    this->m_backgroundColor = newValue;
    this->applyStyleSheet();
}

QColor QSwitchLayer::borderColor() const {
    return this->m_borderColor;
}

void QSwitchLayer::setBorderColor(QColor newValue) {
    this->m_borderColor = newValue;
    this->applyStyleSheet();
}

int QSwitchLayer::borderRadius() const {
    return this->m_borderRadius;
}

void QSwitchLayer::setBorderRadius(int newValue) {
    this->m_borderRadius = newValue;
    this->applyStyleSheet();
}

int QSwitchLayer::borderLineWidth() const {
    return this->m_borderLineWidth;
}

void QSwitchLayer::setBorderLineWidth(int newValue) {
    this->m_borderLineWidth = newValue;
    this->applyStyleSheet();
}

QColor QSwitchLayer::disabledBackgroundColor() const {
    return this->m_disabledBackgroundColor;
}

QColor QSwitchLayer::disabledBorderColor() const {
    return QColor(0, 0, 0, 51);
}

void QSwitchLayer::applyStyleSheet() {
    this->setStyleSheet(
        QSwitchLayer::templateStyleSheet.arg(this->m_backgroundColor.red())
            .arg(this->m_backgroundColor.green())
            .arg(this->m_backgroundColor.blue())
            .arg(this->m_backgroundColor.alpha())
            .arg(this->m_borderLineWidth)
            .arg(this->m_borderColor.red())
            .arg(this->m_borderColor.green())
            .arg(this->m_borderColor.blue())
            .arg(this->m_borderColor.alpha())
            .arg(this->m_borderRadius));
}

QSwitch::QSwitch(QColor tintColor, QWidget *parent)
    : QFrame(parent), m_tintColor(tintColor) {
    this->m_animations = new QParallelAnimationGroup(this);

    this->m_backgroundLayer = new QSwitchLayer(this);
    this->m_backgroundLayer->setObjectName("_backgroundLayer");
    this->m_backgroundLayer->setBorderLineWidth(QSwitch::borderLineWidth);
    this->setMinimumSize(QSwitch::minimumWidth + 2 * QSwitch::defaultMargin,
                         QSwitch::minimumHeight + 2 * QSwitch::defaultMargin);
    this->m_backgroundLayer->move(QSwitch::defaultMargin,
                                  QSwitch::defaultMargin);

    this->m_knobLayer = new QFrame(this);
    this->m_knobLayer->setObjectName("_knobLayer");
    auto knobLayerShadowEffect =
        new QGraphicsDropShadowEffect(this->m_knobLayer);
    knobLayerShadowEffect->setColor(QColor(0, 0, 0, 77));
    knobLayerShadowEffect->setBlurRadius(4);
    knobLayerShadowEffect->setOffset(0, 2);
    this->m_knobLayer->setGraphicsEffect(knobLayerShadowEffect);

    this->m_knobInsideLayer = new QFrame(this);
    this->m_knobInsideLayer->setObjectName("_knobInsideLayer");
    auto knobInsideLayerShadowEffect =
        new QGraphicsDropShadowEffect(this->m_knobInsideLayer);
    knobInsideLayerShadowEffect->setColor(QColor(0, 0, 0, 89));
    knobInsideLayerShadowEffect->setBlurRadius(4);
    knobInsideLayerShadowEffect->setOffset(0, 0);
    this->m_knobInsideLayer->setGraphicsEffect(knobInsideLayerShadowEffect);

    this->redrawLayers();
}

bool QSwitch::isChecked() const {
    return this->m_isChecked;
}

void QSwitch::setChecked(bool newValue) {
    this->m_isChecked = newValue;
    this->redrawLayers(true);
}

QColor QSwitch::tintColor() const {
    return this->m_tintColor;
}

void QSwitch::setTintColor(QColor newValue) {
    this->m_tintColor = newValue;
    this->redrawLayers();
}

void QSwitch::mousePressEvent([[maybe_unused]] QMouseEvent *event) {
    if (!this->isEnabled()) {
        return;
    }
    this->m_isMouseDown = true;
    this->redrawLayers(true);
}

void QSwitch::mouseMoveEvent(QMouseEvent *event) {
    if (!this->m_isMouseDown) {
        return;
    }

    this->m_didDrag = true;

    auto oldIsDraggingTowardsOn = this->m_isDraggingTowardsOn;
    this->m_isDraggingTowardsOn =
        event->pos().x() >=
        static_cast<int>((this->width() - 2 * QSwitch::defaultMargin) / 2.0);

    if (oldIsDraggingTowardsOn != this->m_isDraggingTowardsOn) {
        this->m_isDraggingTowardsOnDidChangeSinceLastMoveEvent = true;
        this->redrawLayers(true);
    }
}

void QSwitch::mouseReleaseEvent([[maybe_unused]] QMouseEvent *event) {
    if (!this->isEnabled()) {
        return;
    }
    this->m_isMouseDown = false;

    if (!this->m_didDrag) {
        this->setChecked(!this->isChecked());
    } else {
        this->setChecked(this->m_isDraggingTowardsOn);
    }

    this->m_didDrag = false;
    this->m_isDraggingTowardsOn = false;

    if (!this->m_isDraggingTowardsOnDidChangeSinceLastMoveEvent) {
        this->redrawLayers(true);
    }

    this->m_isDraggingTowardsOnDidChangeSinceLastMoveEvent = false;
    emit this->toggled(this->isChecked());
}

void QSwitch::resizeEvent(QResizeEvent *event) {
    QFrame::resizeEvent(event);
    this->redrawLayers();
}

QSize QSwitch::sizeHint() const {
    auto hint = this->m_backgroundLayer->sizeHint();
    return QSize(hint.width() + 2 * QSwitch::defaultMargin,
                 hint.height() + 2 * QSwitch::defaultMargin);
}

QEasingCurve QSwitch::makeBezierCurveOvershoot() const {
    auto curve = QEasingCurve(QEasingCurve::BezierSpline);
    curve.addCubicBezierSegment(
        QPointF(0.25, 1.5), QPointF(0.5, 1.0), QPointF(1.0, 1.0));
    return curve;
}

QEasingCurve QSwitch::makeBezierCurveDefault() const {
    auto curve = QEasingCurve(QEasingCurve::BezierSpline);
    curve.addCubicBezierSegment(
        QPointF(0.25, 0.1), QPointF(0.25, 1.0), QPointF(1.0, 1.0));
    return curve;
}

void QSwitch::redrawLayers(bool animated) {
    this->m_backgroundLayer->resize(this->width() - 2 * QSwitch::defaultMargin,
                                    this->height() -
                                        2 * QSwitch::defaultMargin);
    this->m_backgroundLayer->setBorderRadius(
        ((this->height() - 2 * QSwitch::defaultMargin) / 2) - 1);

    this->m_animations->clear();

    if ((this->m_didDrag && this->m_isDraggingTowardsOn) ||
        (!this->m_didDrag && this->isChecked())) {
        auto animationBackgroundColor =
            new QPropertyAnimation(this->m_backgroundLayer, "backgroundColor");
        animationBackgroundColor->setDuration(
            QSwitch::animationDurationMilliSeconds);
        animationBackgroundColor->setStartValue(
            this->m_backgroundLayer->backgroundColor());
        animationBackgroundColor->setEndValue(this->tintColor());
        if (this->m_didDrag) {
            animationBackgroundColor->setEasingCurve(
                this->makeBezierCurveDefault());
        } else {
            animationBackgroundColor->setEasingCurve(
                this->makeBezierCurveOvershoot());
        }
        this->m_animations->addAnimation(animationBackgroundColor);

        auto animationBorderColor =
            new QPropertyAnimation(this->m_backgroundLayer, "borderColor");
        animationBorderColor->setDuration(
            QSwitch::animationDurationMilliSeconds);
        animationBorderColor->setStartValue(
            this->m_backgroundLayer->borderColor());
        animationBorderColor->setEndValue(this->tintColor());
        if (this->m_didDrag) {
            animationBorderColor->setEasingCurve(
                this->makeBezierCurveDefault());
        } else {
            animationBorderColor->setEasingCurve(
                this->makeBezierCurveOvershoot());
        }
        this->m_animations->addAnimation(animationBorderColor);
    } else {
        auto animationBackgroundColor =
            new QPropertyAnimation(this->m_backgroundLayer, "backgroundColor");
        animationBackgroundColor->setDuration(
            QSwitch::animationDurationMilliSeconds);
        animationBackgroundColor->setStartValue(
            this->m_backgroundLayer->backgroundColor());
        animationBackgroundColor->setEndValue(
            this->m_backgroundLayer->disabledBackgroundColor());
        if (this->m_didDrag) {
            animationBackgroundColor->setEasingCurve(
                this->makeBezierCurveDefault());
        } else {
            animationBackgroundColor->setEasingCurve(
                this->makeBezierCurveOvershoot());
        }
        this->m_animations->addAnimation(animationBackgroundColor);

        auto animationBorderColor =
            new QPropertyAnimation(this->m_backgroundLayer, "borderColor");
        animationBorderColor->setDuration(
            QSwitch::animationDurationMilliSeconds);
        animationBorderColor->setStartValue(
            this->m_backgroundLayer->borderColor());
        animationBorderColor->setEndValue(
            this->m_backgroundLayer->disabledBorderColor());
        if (this->m_didDrag) {
            animationBorderColor->setEasingCurve(
                this->makeBezierCurveDefault());
        } else {
            animationBorderColor->setEasingCurve(
                this->makeBezierCurveOvershoot());
        }
        this->m_animations->addAnimation(animationBorderColor);
    }

    auto knobWidth =
        static_cast<int>((this->width() - 2 * QSwitch::defaultMargin -
                          2 * QSwitch::borderLineWidth) *
                         (1 / QSwitch::goldenRatio));
    auto roundKnobWidth = knobWidth;
    if (this->m_isMouseDown) {
        knobWidth =
            static_cast<int>((this->width() - 2 * QSwitch::defaultMargin -
                              2 * QSwitch::borderLineWidth) *
                             (1 / QSwitch::decreasedGoldenRatio));
    }

    if (animated) {
        auto animationSizeKnob =
            new QPropertyAnimation(this->m_knobLayer, "size");
        animationSizeKnob->setDuration(QSwitch::animationDurationMilliSeconds);
        animationSizeKnob->setStartValue(this->m_knobLayer->size());
        animationSizeKnob->setEndValue(
            QSize(knobWidth,
                  this->height() - 2 * QSwitch::defaultMargin -
                      2 * QSwitch::borderLineWidth));
        if (this->m_didDrag) {
            animationSizeKnob->setEasingCurve(this->makeBezierCurveDefault());
        } else {
            animationSizeKnob->setEasingCurve(
                this->makeBezierCurveOvershoot());
        }
        this->m_animations->addAnimation(animationSizeKnob);
    } else {
        this->m_knobLayer->resize(knobWidth,
                                  this->height() - 2 * QSwitch::defaultMargin -
                                      2 * QSwitch::borderLineWidth);
    }

    auto knobX = [this, knobWidth]() -> int {
        if ((!this->m_didDrag && !this->isChecked()) ||
            (this->m_didDrag && !this->m_isDraggingTowardsOn)) {
            return QSwitch::defaultMargin + QSwitch::borderLineWidth;
        } else {
            return this->width() - knobWidth - QSwitch::borderLineWidth -
                   QSwitch::defaultMargin;
        }
    }();

    if (animated) {
        auto animationPosKnob =
            new QPropertyAnimation(this->m_knobLayer, "pos");
        animationPosKnob->setDuration(QSwitch::animationDurationMilliSeconds);
        animationPosKnob->setStartValue(this->m_knobLayer->pos());
        animationPosKnob->setEndValue(
            QPoint(knobX, QSwitch::defaultMargin + QSwitch::borderLineWidth));
        if (this->m_didDrag) {
            animationPosKnob->setEasingCurve(this->makeBezierCurveDefault());
        } else {
            animationPosKnob->setEasingCurve(this->makeBezierCurveOvershoot());
        }
        this->m_animations->addAnimation(animationPosKnob);
    } else {
        this->m_knobLayer->move(
            knobX, QSwitch::defaultMargin + QSwitch::borderLineWidth);
        this->m_knobLayer->setStyleSheet(
            QString("QFrame#_knobLayer { background-color: white; "
                    "border-radius: %1px; }")
                .arg(this->m_knobLayer->height() / 2 - 1));
    }

    if (animated) {
        auto animationSizeKnobInside =
            new QPropertyAnimation(this->m_knobInsideLayer, "size");
        animationSizeKnobInside->setDuration(
            QSwitch::animationDurationMilliSeconds);
        animationSizeKnobInside->setStartValue(
            this->m_knobInsideLayer->size());
        animationSizeKnobInside->setEndValue(
            QSize(knobWidth,
                  this->height() - 2 * QSwitch::defaultMargin -
                      2 * QSwitch::borderLineWidth));
        if (this->m_didDrag) {
            animationSizeKnobInside->setEasingCurve(
                this->makeBezierCurveDefault());
        } else {
            animationSizeKnobInside->setEasingCurve(
                this->makeBezierCurveOvershoot());
        }
        this->m_animations->addAnimation(animationSizeKnobInside);
    } else {
        this->m_knobInsideLayer->resize(knobWidth,
                                        this->height() -
                                            2 * QSwitch::defaultMargin -
                                            2 * QSwitch::borderLineWidth);
    }

    if (animated) {
        auto animationPosKnobInside =
            new QPropertyAnimation(this->m_knobInsideLayer, "pos");
        animationPosKnobInside->setDuration(
            QSwitch::animationDurationMilliSeconds);
        animationPosKnobInside->setStartValue(this->m_knobInsideLayer->pos());
        animationPosKnobInside->setEndValue(
            QPoint(knobX, QSwitch::defaultMargin + QSwitch::borderLineWidth));
        if (this->m_didDrag) {
            animationPosKnobInside->setEasingCurve(
                this->makeBezierCurveDefault());
        } else {
            animationPosKnobInside->setEasingCurve(
                this->makeBezierCurveOvershoot());
        }
        this->m_animations->addAnimation(animationPosKnobInside);
    } else {
        this->m_knobInsideLayer->move(
            knobX, QSwitch::defaultMargin + QSwitch::borderLineWidth);
        this->m_knobInsideLayer->setStyleSheet(
            QString("QFrame#_knobInsideLayer { background-color: white; "
                    "border-radius: %1px; }")
                .arg(this->m_knobInsideLayer->height() / 2 - 1));
    }

    this->m_animations->start();

    this->setMaximumHeight(roundKnobWidth + 2 * QSwitch::defaultMargin +
                           2 * QSwitch::borderLineWidth);
}
